### ISSUE TITLE
Wake Process 3D Potential Flow

### ISSUE THREAD

**Issue by inigolcanalejo** [KratosMultiphysics/Kratos#5278]
**The big question mark is whether it is ok to be using the CalculateDistanceToSkinProcess3D instead of the 
CalculateDiscontinuousDistanceToSkinProcess3D.** I know that @rubenzorrilla and @marcnunezc are recommending to use the discontinuous because it should fulfill more our requirements, however I didn't manage to make it work properly and I will explain why.

Our requirement from the distance process:
- The process should save in each element the ELEMENTAL_DISTANCES so that then we can decide whether the element is cut or not (if the process doesn't do this automatically)

The problem comes for the special cases where there are elements with nodes laying on the wake. The problem here is the definition of cut element. In 2D we decided to move nodes laying on the wake upwards in such a manner that the element on top would not be cut and the element underneath is cut. In the following figure the node marked with the arrow is laying exactly on the wake (or close to the wake) so it is moved upwards so that the elements marked with the green circle are cut.
![node_on_wake](https://user-images.githubusercontent.com/28628414/61526307-7e849300-aa1a-11e9-8824-758057d3eebe.png)

It can also happen that a whole side is laying on the wake and it would work in the same manner. Elements with the circle are cut:
![side_on_wake](https://user-images.githubusercontent.com/28628414/61526619-02d71600-aa1b-11e9-8903-9d752fc4f754.png)

In 3D one can have a node, an edge or a whole face of a tetrahedral laying on the wake. The issue I have with CalculateDiscontinuousDistanceToSkinProcess3D is that it does not detect these special cases in 3D whereas CalculateDistanceToSkinProcess3D does. My question to @marcnunezc : I have noticed that you use https://github.com/KratosMultiphysics/Kratos/blob/f37002e52416991b4e12f71eedff1d5e00fcd1b5/applications/CompressiblePotentialFlowApplication/custom_processes/define_embedded_wake_process.cpp#L48
Does it take care of these special cases in 2D?

**Comment by pooyan-dadvand** [comment]
Are you sure that continuous distance gives better results on this situation? If I remember correctly, @rubenzorrilla and me decided to put all these kind of treatments in separate tools (like embedded_distance) and keep the calculate distances as pure as possible without any special modification. @rubenzorrilla isn't it?

**Comment by marcnunezc** [comment]
Hi I'd like to add some examples in 2D (which are easier to visualize), that might help us understanding the proper way of using these tools.

Here (#5261) I am using the 2D Discontinuous process  as @inigolcanalejo  is pointing out to find the elements that are cut by a line (which is the wake). In general, I have seen that the Discontinuous process does the job:

![image](https://user-images.githubusercontent.com/32136457/61638768-5efb9d80-ac9a-11e9-81b6-db46acb12e80.png)

However, the issue comes when I force the wake to pass to exactly through one node (where the distance is 0.0). I know this is highly unlikely to happen in an unstructured mesh, but in case it does, the search of the "WAKE" elements goes wrong. Here I am showing the values of the TO_SPLIT flag that come from the DiscontinuousDistanceProcess:
![image](https://user-images.githubusercontent.com/32136457/61640777-37a6cf80-ac9e-11e9-8418-1923d5f42f69.png)
![image](https://user-images.githubusercontent.com/32136457/61640957-818fb580-ac9e-11e9-9371-cff3ba74a4f7.png)
Note that the node that the wake is intersecting has DISTANCE=0.0, and a few elements that should be marked, aren't.

Now if the **continuous** distance is used, the TO_SPLIT flag is correcly set:

![image](https://user-images.githubusercontent.com/32136457/61641162-ef3be180-ac9e-11e9-9eb9-418141ffa60a.png)

![image](https://user-images.githubusercontent.com/32136457/61641279-23af9d80-ac9f-11e9-8b34-cd2cafdd786e.png)



Where the node which  is interesected by the wake now has some **epsilon** instead of 0.0.

I would say that to us the "0.0" is not the problem, but to have the TO_SPLIT flag undefined in the elements close to the 0.0 distance.

If the TO_SPLIT elements were the same in the discontinuous approach as the continuous approach it would be fine. Also i think this behaviour is what @inigolcanalejo was mentioning when he said that the continuous process works better for this special cases.


> Are you sure that continuous distance gives better results on this situation? If I remember correctly, @rubenzorrilla and me decided to put all these kind of treatments in separate tools (like embedded_distance) and keep the calculate distances as pure as possible without any special modification. @rubenzorrilla isn't it?

Regarding this, is the DiscontinousProcess **not meant** to deal with these special cases (and rather this should be "postprocessed"?)?

**Comment by pooyan-dadvand** [comment]
> Regarding this, is the DiscontinousProcess not meant to deal with these special cases (and rather this should be "postprocessed"?)?

Yes, For the same reason that on that time @RiccardoRossi had different special cases for wake than the ones of @rubenzorrilla for embedded we decided to do it differently afterwards.

**Comment by rubenzorrilla** [comment]
First of all, and as we have previously discussed, the point in here is not which process gives you the best distance according to your requirements. The point is that volumeless geometries, as your line or plane wake, can only be represented by a discontinuous level set (`CalculateDiscontinuousDistanceToSkinProcess`), since the inside / outside concept associated to a continuous level set is a complete non-sense for such geometries.

Having said this, I think that your test cases are working because the wake you are setting is completely intersecting the domain, so one half of the domain is the positive volume and the other one is the negative. This wouldn't work for a real wake test.

Thus, my advise is to completely forget the continuous distance proces (`CalculateDistanceToSkinProcess`) and to focus on why the discontinuous one doesn't yield the expected behaviour. In my opinion, such limit case leads to a tolerances problem when computing the intersections, and this is why the elements appear as non-split. As @pooyan-dadvand has pointed out above, what we do in the discontinuous process is a "naive" distance calculation, which doesn't do any special treatment of these cases.

To try to answer why the continuous one is working, I guess that it takes the distance value from the neighbor while doing the nodal smoothing of the discontinuous one. In any case, the value you get is something up to the -16, which is actually in the order of the machine precision. Thus, I wouldn't rely so much on it.

Long story short, I would completely forget about the continuous distance in this case and isolate one of the "ill" elements in your large test to see what can we do to consider these cases. However, I don't think we should modify the distance algorithm, as it is properly performing (you get zero distance since the node is lying on the wake). Maybe we can thing on an optional pre-processing to tackle these cases.

**Comment by inigolcanalejo** [comment]
Hi, thanks everyone for all you comments. @marcnunezc you illustrated exactly what I meant with your example. 

> If the TO_SPLIT elements were the same in the discontinuous approach as the continuous approach it would be fine.

This is exactly the point. Let me also just make a few notes:

> I know this is highly unlikely to happen in an unstructured mesh

From my experience this actually does happen quite often so I think we need to find a robust solution for this.

I also must say that I fully agree with @rubenzorrilla that it doesn't make sense to represent volumeless geometries by a discontinuous level set. However, I think I didn't explain myself correctly and I believe that the source of the confusion is that both CalculateDiscontinuousDistanceToSkinProcess and CaltulateDistanceToSkinProcess compute a discontinuous distance and they do it differently.  Let me try to link to the code so that you know what I mean. Since you probably know better than me the [CaltulateDistanceToSkinProcess is a derived class from CalculateDiscontinuousDistanceToSkinProcess](https://github.com/KratosMultiphysics/Kratos/blob/dc0d9e31bec2b685e534848dad108d5fe75621be/kratos/processes/calculate_distance_to_skin_process.h#L38). CaltulateDistanceToSkinProcess computes [both](https://github.com/KratosMultiphysics/Kratos/blob/dc0d9e31bec2b685e534848dad108d5fe75621be/kratos/processes/calculate_distance_to_skin_process.cpp#L75-L90):

- [a continuous distance](https://github.com/KratosMultiphysics/Kratos/blob/dc0d9e31bec2b685e534848dad108d5fe75621be/kratos/processes/calculate_distance_to_skin_process.cpp#L88) stored in the variable DISTANCE (which I fully agree with you that does not make any sense in our case) 
- and it also computes [a discontinuous distance](https://github.com/KratosMultiphysics/Kratos/blob/dc0d9e31bec2b685e534848dad108d5fe75621be/kratos/processes/calculate_distance_to_skin_process.cpp#L85) stored in the variable ELEMENTAL_DISTANCES (this is the one we want!). 

The point here is that the [discontinuous distance is computed differently](https://github.com/KratosMultiphysics/Kratos/blob/dc0d9e31bec2b685e534848dad108d5fe75621be/kratos/processes/calculate_distance_to_skin_process.cpp#L78-L86) in the [CaltulateDistanceToSkinProcess](https://github.com/KratosMultiphysics/Kratos/blob/dc0d9e31bec2b685e534848dad108d5fe75621be/kratos/processes/calculate_distance_to_skin_process.cpp#L94-L134) than in the [CalculateDiscontinuousDistanceToSkinProcess](https://github.com/KratosMultiphysics/Kratos/blob/dc0d9e31bec2b685e534848dad108d5fe75621be/kratos/processes/calculate_discontinuous_distance_to_skin_process.cpp#L118-L187). And it just turns out that the way the discontinuous distance is compute in CaltulateDistanceToSkinProcess is more robust for our case [because it takes into account the case where nodes from the volume mesh are laying on the skin](https://github.com/KratosMultiphysics/Kratos/blob/dc0d9e31bec2b685e534848dad108d5fe75621be/kratos/processes/calculate_distance_to_skin_process.cpp#L154). This doesn't mean we are using a continuous distance, note we are still using a discontinuous distance!

Answering your points:

> Having said this, I think that your test cases are working because the wake you are setting is completely intersecting the domain, so one half of the domain is the positive volume and the other one is the negative. This wouldn't work for a real wake test.

Actually I also tested it for cases where the wake does not intersect the domain entirely and real wake tests and it also works. Find example below.

> To try to answer why the continuous one is working, I guess that it takes the distance value from the neighbor while doing the nodal smoothing of the discontinuous one. In any case, the value you get is something up to the -16, which is actually in the order of the machine precision. Thus, I wouldn't rely so much on it.

I insist that we are only looking at the discontinuous distance stored in ELEMENTAL_DISTANCES and not at the continuous distance stored in DISTANCE (as it doesn't make sense to look at the continuous distance in this case as you pointed out). Regarding the small values of the order of magnitude of 10⁻¹⁶, this result is expected since in the computation of the discontinuous distance within the CaltulateDistanceToSkinProcess nodes laying on the wake are moved downwards a given epsilon. This is done here:

https://github.com/KratosMultiphysics/Kratos/blob/dc0d9e31bec2b685e534848dad108d5fe75621be/kratos/processes/calculate_distance_to_skin_process.cpp#L153-L154

Where Epsilon is defined here:

https://github.com/KratosMultiphysics/Kratos/blob/dc0d9e31bec2b685e534848dad108d5fe75621be/kratos/processes/calculate_distance_to_skin_process.cpp#L110

> Long story short, I would completely forget about the continuous distance in this case and isolate one of the "ill" elements in your large test to see what can we do to consider these cases. However, I don't think we should modify the distance algorithm, as it is properly performing (you get zero distance since the node is lying on the wake). Maybe we can thing on an optional pre-processing to tackle these cases.

I agree with you that we need to forget about the continuous distance and that we should not modify the distance algorithms as they are properly performing. I would just suggest using the discontinuous distance computed in the CaltulateDistanceToSkinProcess since it already takes care of all the special cases.

Regarding the core, just for your information, @pooyan-dadvand and @rubenzorrilla I don't know if you are aware of the fact that in the CaltulateDistanceToSkinProcess there is a hard coded flag choosing always the so called "naive elemental distance computation (without plane optimization)" (which by the way, it is the one that is taking care of the special cases in our examples), whereas the elemental distances computed as in the parent class is not used:

https://github.com/KratosMultiphysics/Kratos/blob/dc0d9e31bec2b685e534848dad108d5fe75621be/kratos/processes/calculate_distance_to_skin_process.cpp#L78-L86

I hope I managed to make more clear what I meant. Thanks a lot to all of you for your patience!

Example where the wake is not fully cutting the entire domain:
![wake](https://user-images.githubusercontent.com/28628414/61697371-6c1a9a00-ad37-11e9-8107-fd2dc05c7a18.png)

The wake elements as expected (using the discontinuous distance computation of the CaltulateDistanceToSkinProcess):
![wake_elements](https://user-images.githubusercontent.com/28628414/61697408-82285a80-ad37-11e9-8688-4e8b2366c787.png)

The outcome of the CalculateDiscontinuousDistanceToSkinProcess where wake elements are missing.
![wake_elements_discontinuous](https://user-images.githubusercontent.com/28628414/61697483-a1bf8300-ad37-11e9-8040-66269f4824b1.png)

**Comment by rubenzorrilla** [comment]
This clarifies things a lot. You are using the discontinuous level set that is computed before the nodal smoothing and raycasting.

**Comment by marcnunezc** [comment]
Thanks @inigolcanalejo for your detailed explanation!

I would say that we can conclude with the following questions:

- Is it your plan to add the same ElementalDistance calculation that is present in the **continuous** distance process to the Discontinuous process?

- In case it is not, is it fine for you if we use the **continuous** process to retrieve the elemental distance in our application?

Regards,

**Comment by inigolcanalejo** [comment]
Yes, it is fine with me to use the continuous process to retrieve the elemental distance, I actually wanted to suggest that too. I think right now that is the most straightforward and robust approach.

> This clarifies things a lot. You are using the discontinuous level set that is computed before the nodal smoothing and raycasting.

Exactly!

**Comment by pooyan-dadvand** [comment]
> Regarding the core, just for your information, @pooyan-dadvand and @rubenzorrilla I don't know if you are aware of the fact that in the CaltulateDistanceToSkinProcess there is a hard coded flag choosing always the so called "naive elemental distance computation (without plane optimization)" (which by the way, it is the one that is taking care of the special cases in our examples), whereas the elemental distances computed as in the parent class is not used:

This was added by @rubenzorrilla

**Comment by rubenzorrilla** [comment]
> > Regarding the core, just for your information, @pooyan-dadvand and @rubenzorrilla I don't know if you are aware of the fact that in the CaltulateDistanceToSkinProcess there is a hard coded flag choosing always the so called "naive elemental distance computation (without plane optimization)" (which by the way, it is the one that is taking care of the special cases in our examples), whereas the elemental distances computed as in the parent class is not used:
> 
> This was added by @rubenzorrilla

Yes, when we decided to enhance the discontinuous level set by using the plane optimization we also decided to keep the continuous level set with the simple distance calculation. This was done for the sake of computational efficiency, relying on the nodal smoothing after the elemental distance calculation. Maybe we can expose the flag to the user.

In any case, can you prepare a small (as small as you can) test case to be added to the core suite. With this we can better take a decision about this.

I don't think the way to go is to use the discontinuous level-set and then do a nodal smoothing in your custom potential flow utilities.

**Comment by philbucher** [comment]
is this still active?

**Comment by inigolcanalejo** [comment]
As far as I know it is still work in progress. The discussion is not active right now, but I think it is not concluded.

**Comment by marcnunezc** [comment]
Hi!

Yes sorry, this is still active. 

I have to update the PR in #5595 and ask again for review, which should fix this issue.
